src/librustc_passes/rvalue_promotion.rs

```diff
@@ -139,19 +139,26 @@ impl BitAnd for Promotability {
 
 impl<'a, 'gcx> CheckCrateVisitor<'a, 'gcx> {
     // Returns true iff all the values of the type are promotable.
-    fn type_has_only_promotable_values(&mut self, ty: Ty<'gcx>) -> bool {
-        debug!("In type_has_only_promotable_values(), ty: {:#?}", ty);
+    fn type_promotability(&mut self, ty: Ty<'gcx>) -> Promotability {
+        debug!("type_promotability({})", ty);
 
-        ty.is_freeze(self.tcx, self.param_env, DUMMY_SP) &&
-            !ty.needs_drop(self.tcx, self.param_env)
+        if ty.is_freeze(self.tcx, self.param_env, DUMMY_SP) &&
+            !ty.needs_drop(self.tcx, self.param_env) {
+            Promotable
+        } else {
+            NotPromotable
+        }
     }
 
     fn handle_const_fn_call(&mut self, def_id: DefId,
                             ret_ty: Ty<'gcx>, span: Span) -> Promotability {
         debug!("In handle_const_fn_call(), def_id: {:#?}, span: {:#?}, ret_ty: {:#?}",
                def_id, span, ret_ty);
 
-        if !self.type_has_only_promotable_values(ret_ty) { return NotPromotable; }
+        let promotability_output = self.type_promotability(ret_ty);
+        if let Promotable = promotability_output {
```

Comment:
remove this `if let` and just `&` the `promotability_output` into the other variables at the end of this function